expertsystem/state/particle/__init__.py

```diff
@@ -0,0 +1,437 @@
+
+    def __float__(self) -> float:
+        return self.__magnitude
+
+    @property
+    def magnitude(self) -> float:
+        return self.__magnitude
+
+    @property
+    def projection(self) -> float:
+        return self.__projection
+
+    def __repr__(self) -> str:
+        return f"(mag: {self.__magnitude}, proj: {self.__projection})"
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, Spin):
+            return (
+                self.__magnitude == other.magnitude
+                and self.__projection == other.projection
+            )
+        if isinstance(other, (float, int)):
+            return self.__magnitude == float(other)
+        raise NotImplementedError
+
+    def __hash__(self) -> Any:
+        return hash(repr(self))
+
+
+class Parity:
+    """Restrict values to 𝕫₂."""
+
+    def __init__(self, parity: Union[float, int, str] = 0) -> None:
+        if isinstance(parity, str):
+            if parity in ["odd", "-1"]:
+                parity = -1
+            elif parity in ["even", "1", "+1"]:
+                parity = +1
+            else:
+                parity = 0
+        if isinstance(parity, float):
+            if parity.is_integer():
+                parity = int(parity)
+        if parity not in [-1, +1, 0]:
+            raise ValueError("Parity can only be -1 or +1")
+        self.__value = int(parity)
+
+    def __int__(self) -> int:
+        return self.__value
+
+    def __bool__(self) -> bool:
+        return self.__value != 0
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, Parity):
+            return self.__value == int(other)
+        return int(self) == other
+
+    def __repr__(self) -> str:
+        output = "Parity: "
+        if self.__value == +1:
+            return output + "+1"
+        if self.__value == 0:
+            return output + "UNDEFINED"
+        return output + str(self.__value)
+
+
+class Parameter(NamedTuple):  # noqa: D101
+    value: float = 0.0
+    uncertainty: float = 0.0
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, Parameter):
+            return (
+                self.value == other.value
+                and self.uncertainty == other.uncertainty
+            )
+        return self.value == other
+
+    def __float__(self) -> float:
+        return self.value
+
+
+class Particle(NamedTuple):  # noqa: D101
+    name: str
+    pid: int
+    mass: Parameter
+    charge: float
+    spin: Spin
+    width: Optional[Parameter] = None
+    isospin: Spin = Spin()
+    parity: Parity = Parity()
+    parity_c: Parity = Parity()
+    parity_g: Parity = Parity()
+    strangeness: int = 0
+    charmness: int = 0
+    bottomness: int = 0
+    topness: int = 0
+    baryon_number: int = 0
+    ln_electron: int = 0
+    ln_muon: int = 0
+    ln_tau: int = 0
+
+    @property
+    def has_width(self) -> bool:
+        return self.width is not None
+
+
+class ParticleDatabase:
```

Comment:
This looks quite big for what it can do. I have to take a closer look at this

Reply:
Yes, the 'problem' is the functionality to load from both XML and YAML (and, worse, the existing XML structure). Technically, this class contains several factories to construct `Particle` objects and the objects it contains from `dict` structures, such as `load_xml`. The interface doesn't reveal that because these factories are wrapped as inline functions or private methods in case the dict structure is the comparable in the case of both XML and YAML.

Reply:
I start to think this problem is strongly related to #22. In the end, this module should just serve as a collection of data containers for particle info -- the `load` and `write` methods are related to serialisation to a custom XML or YAML format that not only contains particle info, but also the intensity recipe etc.

So what I'm currently giving a try is to design some sort of file adapter or recipe module that can be easily expanded, but that for now would only serve as a factory for the `ParticleDatabase` class. This module should also easily be accesible usable by an external package (i.e. Tensorwaves).

Reply:
Okay, let's move this discussion to #44, it's a bigger design decision. The current PR should only fix #42, so it was a mistake to start working on `load`/`write` functions in this PR already.